.github/workflows/ci.yml

```diff
@@ -26,13 +26,54 @@ concurrency:
   cancel-in-progress: true
 
 jobs:
+  build-linter:
+    name: Build Custom Linter
+    runs-on: ubuntu-latest
+    steps:
+      - name: Check out code
+        uses: actions/checkout@v4
+      - name: Set up Go
+        uses: actions/setup-go@v5
+        with:
+          go-version: ${{ env.GO_VERSION }}
+          cache: true
+      - name: Cache custom linter binary
+        id: cache-linter
+        uses: actions/cache@v3
+        with:
+          # Key should change whenever implementation (tools/structwrite), or compilation config (.custom-gcl.yml) changes
+          # When the key is different, it is a cache miss, and the custom linter binary is recompiled
+          key: custom-linter-${{ env.GO_VERSION }}-${{ runner.os }}-${{ hashFiles('.custom-gcl.yml', 'tools/structwrite/**') }}
+          path: tools/custom-gcl # path defined in .custom-gcl.yml
+      # We install the non-custom golangci-lint binary using the golangci-lint action.
+      # The action is set up to always install and run the linter - there isn't a way to only install.
+      # We provide args to disable all linters which results in the linter run immediately failing.
+      - name: Install golangci-lint
+        if: steps.cache-linter.outputs.cache-hit != 'true'
+        uses: golangci/golangci-lint-action@v6
+        continue-on-error: true # after installation (what we care about), the run will fail - this line allows workflow to continue
+        with:
+          # Required: the version of golangci-lint is required and must be specified without patch version: we always use the latest patch version.
+          version: v1.63
+          args: "--no-config --disable-all" # set args so that no linters are actually run
+      - name: Build custom linter binary
+        if: steps.cache-linter.outputs.cache-hit != 'true'
+        run: |
+          golangci-lint custom
+      - name: Upload linter binary as artifact
+        uses: actions/upload-artifact@v4
+        with:
+          name: custom-gcl
+          path: tools/custom-gcl # path defined in .custom-gcl.yml
+
```

Comment:
Does this work on subsequent runs? Based on the documentation of the [upload-artifact action](https://github.com/actions/upload-artifact?tab=readme-ov-file#overwriting-an-artifact), this is immutable, and I would expect this to fail. Unless we change the `name` to be dynamic, I would expect this to fail on subsequent runs.

Along the same lines, can we share the binary across jobs using the cache rather than uploading on every execution? This would not only remove the need for uploads, but it would also ensure that we are are not creating unnecessary spend. For context, I believe that we only have a couple GBs of storage for artifacts. This binary is 14.7MB. If we have to upload it, we would want to modify this to ensure that it is only upload if the file is changed.

Reply:
It did work on subsequent runs with artifacts. However, I have changed the workflow to only use the cache to share the binary between jobs like you suggested: https://github.com/onflow/flow-go/pull/7322/commits/809ded1b9cbb92092e675f14db18546df5d4c04c. Tested on the final few commits of https://github.com/jordanschalm/flow-go/pull/1. 

Using the cache directly seems like a better approach, especially if we're going to blow past our artifact storage allowance. I guess there is the risk of failure due to cache eviction, but it seems very unlikely given that we are writing immediately before use.